lib/northbound_cli.c

```diff
@@ -1423,45 +1425,65 @@ DEFPY (show_yang_operational_data,
        "JavaScript Object Notation\n"
        "Extensible Markup Language\n"
        "Translate operational data\n"
-       "YANG module translator\n")
+       "YANG module translator\n"
+       "Maximum number of elements to fetch at once\n"
+       "Maximum number of elements to fetch at once\n"
+       "Fetch all data using batches of the provided size\n")
 {
+	struct nb_oper_data_iter_input iter_input = {};
+	struct nb_oper_data_iter_output iter_output = {};
 	LYD_FORMAT format;
-	struct yang_translator *translator = NULL;
 	struct ly_ctx *ly_ctx;
 	struct lyd_node *dnode;
 	char *strp;
+	int ret;
 
 	if (xml)
 		format = LYD_XML;
 	else
 		format = LYD_JSON;
 
 	if (translator_family) {
-		translator = yang_translator_find(translator_family);
-		if (!translator) {
+		iter_input.translator = yang_translator_find(translator_family);
+		if (!iter_input.translator) {
 			vty_out(vty, "%% Module translator \"%s\" not found\n",
 				translator_family);
 			return CMD_WARNING;
 		}
 
-		ly_ctx = translator->ly_ctx;
+		ly_ctx = iter_input.translator->ly_ctx;
 	} else
 		ly_ctx = ly_native_ctx;
 
 	/* Obtain data. */
 	dnode = yang_dnode_new(ly_ctx, false);
-	if (nb_oper_data_iterate(xpath, translator, 0, nb_cli_oper_data_cb,
-				 dnode)
-	    != NB_OK) {
-		vty_out(vty, "%% Failed to fetch operational data.\n");
+	iter_input.xpath = xpath;
+	iter_input.cb = nb_cli_oper_data_cb;
+	iter_input.cb_arg = dnode;
+	if (max_elements_str)
+		iter_input.max_elements = max_elements;
+
+	for (;;) {
+		ret = nb_oper_data_iterate(&iter_input, &iter_output);
+		if (!repeat || ret != NB_ITER_STOP)
+			break;
```

Comment:
I'm a little confused here. It seems like it should be `ret != NB_ITER_CONTINUE` or `ret == NB_ITER_STOP` for our loop-break condition? If it's `CONTINUE` don't we want to proceed to update the offset path with where we stopped, as below, and continue looping?

Reply:
`NB_ITER_STOP` is returned when `nb_oper_data_iterate()` has more data to iterate over but had to stop due to the provided maximum chunk size. In the case of the other return values (`NB_ITER_FINISH` and `NB_ITER_ABORT`) there's no need to do another iteration round.

The `NB_ITER_CONTINUE` return value is only used internally and can't be returned by `nb_oper_data_iterate()`. I've fixed the documentation comment of that function to make that clear.

If you prefer, I can also rename `NB_ITER_STOP` to `NB_ITER_SUSPEND` as "suspend" implies that the iteration needs to be resumed at one point, while "stop" doesn't give any hint in that direction.

Reply:
@qlyoung just pushed another update including the `s/NB_ITER_STOP/NB_ITER_SUSPEND/` rename as per our conversation.